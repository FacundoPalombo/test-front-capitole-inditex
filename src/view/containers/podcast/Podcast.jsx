import React from 'react';
import { getPodcastEpisodes } from '../../../services/podcast';
import { Outlet, useLoaderData, useParams, Link } from 'react-router-dom';
import PodcastDetail from './components/PodcastDetail';
import styles from './styles.module.scss';
import { getPodcastChannels } from '../../../services/podcast';

const Podcasts = () => {
  const { podcasts, channels } = useLoaderData();
  const podcast = podcasts.results.find((p) => p.kind === 'podcast');
  const params = useParams();

  /* Quite a bit overkill all this implementation because
    description is on a different api call. But there it is.*/
  const description = channels.feed.entry.find(
    (channel) => channel.id.attributes['im:id'] === params.podcastId
  ).summary.label;

  return (
    <section id="podcasts" className={styles.podcasts}>
      <div data-testid="podcast-detail" className={styles.podcasts__page}>
        <section className={styles.podcasts__detail}>
          <Link
            to={`/podcast/${params.podcastId}`}
            className={styles.podcasts__detail_navigable}
          >
            <PodcastDetail
              image={podcast.artworkUrl600}
              title={podcast.trackName}
              artist={podcast.artistName}
              description={description}
            />
          </Link>
        </section>
        <Outlet />
      </div>
    </section>
  );
};

export async function loader({ params }) {
  const [podcasts, channels] = await Promise.all([
    getPodcastEpisodes({ podcastId: params.podcastId }),
    getPodcastChannels(),
  ]);

  return { podcasts, channels };
}

export default Podcasts;
