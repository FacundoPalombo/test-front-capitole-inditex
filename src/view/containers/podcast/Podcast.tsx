import React from 'react';
import { Outlet, useParams } from 'react-router-dom';
import PodcastDetail from './components/PodcastDetail/PodcastDetail';
import styles from './styles.module.scss';
import {
  episodes as episodesQuery,
  channels as channelsQuery,
} from '@App/queries/podcasts';
import { useIsFetching, useQuery } from '@tanstack/react-query';
import DetailSkeleton from '@components/DetailSkeleton/DetailSkeleton';

const Podcasts = () => {
  const params = useParams();
  const { data: episodes } = useQuery(episodesQuery(params.podcastId));
  const { data: channels } = useQuery(channelsQuery());
  const isLoading = useIsFetching() > 0;

  const episode = episodes?.results?.find((p) => p.kind === 'podcast');

  const description = channels?.feed.entry.find(
    (channel) => channel.id.attributes['im:id'] === params.podcastId
  )?.summary.label;

  // skeleton for early feedback
  if (isLoading) return <DetailSkeleton />;
  return (
    <section id="podcasts" className={styles.podcasts}>
      <div data-testid="podcast-detail" className={styles.podcasts__page}>
        {episode && (
          <PodcastDetail
            image={episode.artworkUrl600!}
            title={episode.trackName!}
            artist={episode.artistName!}
            description={description!}
          />
        )}
        <Outlet />
      </div>
    </section>
  );
};

export const loader =
  (queryClient) =>
  async ({ params }) => {
    const episodesResolvedQuery = episodesQuery(params.podcastId);
    const channelsResolvedQuery = channelsQuery();
    const episodes = await queryClient.ensureQueryData(episodesResolvedQuery);
    const channels = await queryClient.ensureQueryData(channelsResolvedQuery);
    return { channels, episodes };
  };

export default Podcasts;
