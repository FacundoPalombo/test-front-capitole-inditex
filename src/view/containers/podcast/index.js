export { default } from './Podcast';

export { loader } from './Podcast';
