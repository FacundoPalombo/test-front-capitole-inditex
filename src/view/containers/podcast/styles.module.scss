.podcasts {
  margin: 0 16px;

  &__page {
    display: flex;
    flex-direction: row;
    gap: 72px;
    justify-content: stretch;
    align-items: flex-start;
    width: 100%;
    height: 100%;
  }
  &__detail {
    width: auto;
    height: auto;
    display: inline-block;
    box-sizing: content-box;
    &_navigable {
      text-decoration: none;
      color: inherit;
      &:visited {
        color: inherit;
      }
    }
  }
}
