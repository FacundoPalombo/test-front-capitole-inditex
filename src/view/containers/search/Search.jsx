import React, { useState } from 'react';
import { useLoaderData, Await, defer } from 'react-router-dom';
import { getSongs } from '../../../services/songs';
import Song from './components/Song';
import styles from './styles.module.scss';
import SearchBox from './components/SearchBox';

const Search = () => {
  const { songs } = useLoaderData();
  const [searchValue, setSearchValue] = useState('');

  function renderComponent(songs) {
    return (
      <section id="search" className={styles.search__container}>
        <SearchBox
          onChange={(e) => setSearchValue(e.target.value)}
          value={searchValue}
        />
        {searchValue.length > 0
          ? songs.feed.entry
              .filter(
                (song) =>
                  song['im:artist']?.label
                    .toLowerCase()
                    ?.includes(searchValue.toLowerCase().trim()) ||
                  song['im:name']?.label
                    .toLowerCase()
                    ?.includes(searchValue.toLowerCase().trim())
              )
              .map((song) => (
                <Song
                  key={song.id.attributes['im:id']}
                  podcastId={song.id.attributes['im:id']}
                  image={song['im:image'].at(0).label}
                  artist={song['im:artist'].label}
                  title={song['im:name'].label}
                />
              ))
          : songs.feed.entry.map((song) => (
              <Song
                key={song.id.attributes['im:id']}
                podcastId={song.id.attributes['im:id']}
                image={song['im:image'].at(0).label}
                artist={song['im:artist'].label}
                title={song['im:name'].label}
              />
            ))}
      </section>
    );
  }
  return (
    <React.Suspense fallback={<section>loading</section>}>
      <Await
        resolve={songs}
        errorElement={<div>Ups! An errror ocurred fetching songs</div>}
      >
        {renderComponent}
      </Await>
    </React.Suspense>
  );
};

export function loader() {
  const songs = getSongs();
  return defer({ songs });
}

export default Search;
