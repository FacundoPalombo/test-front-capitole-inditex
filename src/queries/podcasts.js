import { getPodcastChannels, getPodcastEpisodes } from '../services/podcast';
import { DAY_IN_MILLISECONDS } from '../utils/constants/various';

export const episodes =
  (queryClient) =>
  async ({ podcastId }) => {
    const episodesInCache = await queryClient.getQueryData(
      `podcasts/${podcastId}`
    );
    let episodes = episodesInCache;

    if (!episodesInCache) {
      episodes = await queryClient.fetchQuery(
        `podcasts/${podcastId}`,
        async () => await getPodcastEpisodes({ podcastId: podcastId }),
        { staleTime: DAY_IN_MILLISECONDS }
      );
    }
    return episodes;
  };

export const channels = (queryClient) => async () => {
  const channelsInCache = await queryClient.getQueryData('channels');
  let channels = channelsInCache;

  if (!channelsInCache) {
    channels = await queryClient.fetchQuery(
      'channels',
      async () => await getPodcastChannels(),
      { staleTime: DAY_IN_MILLISECONDS }
    );
  }
  return channels;
};
