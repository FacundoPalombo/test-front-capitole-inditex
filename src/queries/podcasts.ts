import { getPodcastChannels, getPodcastEpisodes } from '../services/podcasts';

export const episodes = (podcastId) => ({
  queryKey: ['podcasts', podcastId],
  queryFn: () => getPodcastEpisodes({ podcastId }),
});

export const channels = () => ({
  queryKey: ['channels'],
  queryFn: getPodcastChannels,
});
