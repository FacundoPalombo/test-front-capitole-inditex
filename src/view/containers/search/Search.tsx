import React, { useCallback, useState } from 'react';
import Channel from './components/Channel/Channel';
import styles from './styles.module.scss';
import SearchBox from './components/SearchBox/SearchBox';
import { channels as channelsQuery } from '@App/queries/podcasts';
import { useIsFetching, useQuery } from '@tanstack/react-query';
import SearchSkeleton from '@components/SearchSkeleton/SearchSkeleton';

const Search = () => {
  const { data: channels } = useQuery(channelsQuery());
  const [searchValue, setSearchValue] = useState('');
  const handleSearchBox = useCallback(
    (e) => setSearchValue(e.target.value),
    [searchValue]
  );

  const isLoading = useIsFetching() > 0;

  const channelsFiltered = channels?.feed?.entry?.filter(
    (channel) =>
      channel['im:artist'].label
        ?.toLowerCase()
        .includes(searchValue.toLowerCase().trim())! ||
      channel['im:name']?.label
        ?.toLowerCase()
        .includes(searchValue.toLowerCase().trim())
  );

  const resultsCount = channelsFiltered?.length;

  const renderChannels = () =>
    channelsFiltered?.map((channel) => {
      const podcastId = channel?.id?.attributes['im:id'];
      const image = channel['im:image']?.at(-1)?.label!;
      const artist = channel['im:artist'].label;
      const title = channel['im:name']?.label;
      return (
        <Channel
          key={podcastId}
          podcastId={podcastId}
          image={image}
          artist={artist}
          title={title}
        />
      );
    });

  if (isLoading) return <SearchSkeleton />;
  return (
    channels && (
      <section
        id="search"
        data-testid="search"
        className={styles.search__container}
      >
        <SearchBox
          onChange={handleSearchBox}
          value={searchValue}
          resultsCount={resultsCount!}
        />
        {renderChannels()}
      </section>
    )
  );
};

export const loader = (queryClient) => async () => {
  const query = channelsQuery();
  const channels = await queryClient.ensureQueryData(query);
  return channels;
};

export default Search;
