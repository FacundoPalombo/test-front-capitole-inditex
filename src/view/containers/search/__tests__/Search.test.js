import React from 'react';
import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import dataFixture from './podcastChannels.fixture';
import { channels as channelsQuery } from '../../../../queries/podcasts';
import Search, { loader as searchLoader } from '../Search';
import { getPodcastChannels } from '../../../../services/podcast';

jest.mock('../components/Channel', () => (props) => (
  <div>Channel{JSON.stringify(props)}</div>
));
jest.mock('../components/SearchBox', () => (props) => (
  <div>SearchBox {JSON.stringify(props)}</div>
));

jest.mock('../../../components/SearchSkeleton', () => () => (
  <div>SearchSkeleton</div>
));

jest.mock('../../../../services/podcast');

describe('Search unit test', () => {
  it('should match snapshot when isFetching shows skeleton', () => {
    const queryClient = new QueryClient({
      logger: {
        log: console.log,
        warn: console.warn,
        error: () => {},
      },
    });
    getPodcastChannels.mockResolvedValue(dataFixture);
    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <Search />
      </QueryClientProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('should match snapshot when data is provided', async () => {
    const queryClient = new QueryClient({
      logger: {
        log: console.log,
        warn: console.warn,
        error: () => {},
      },
    });
    getPodcastChannels.mockResolvedValue(dataFixture);
    const { asFragment, findByText } = render(
      <QueryClientProvider client={queryClient}>
        <Search />
      </QueryClientProvider>
    );
    await findByText(`SearchBox {"value":"","resultsCount":9}`);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should resolve loader as pure function', async () => {
    const queryClientStub = {
      ensureQueryData: jest.fn().mockImplementationOnce((query) => {
        expect(query).toEqual(channelsQuery());
        return { data: 'ok' };
      }),
    };

    const loader = await searchLoader(queryClientStub)();
    expect(loader).toEqual({ data: 'ok' });
  });
});
