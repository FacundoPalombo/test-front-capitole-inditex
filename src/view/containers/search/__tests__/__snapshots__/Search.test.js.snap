// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search unit test should match snapshot when data is provided 1`] = `
<DocumentFragment>
  <section
    class="search__container"
    data-testid="search"
    id="search"
  >
    <div>
      SearchBox {"value":"","resultsCount":9}
    </div>
    <div>
      Channel{"podcastId":"1535809341","image":"https://is1-ssl.mzstatic.com/image/thumb/Podcasts113/v4/f2/21/fa/f221fabd-017f-5125-633b-f1fe4f39802a/mza_182995249085044287.jpg/170x170bb.png","artist":"The Joe Budden Network","title":"The Joe Budden Podcast"}
    </div>
    <div>
      Channel{"podcastId":"1574029840","image":"https://is5-ssl.mzstatic.com/image/thumb/Podcasts116/v4/5b/88/a1/5b88a186-ce6a-5268-cc49-8e896c737729/mza_7354436562524572096.jpg/170x170bb.png","artist":"Friday Night Karaoke","title":"Friday Night Karaoke"}
    </div>
    <div>
      Channel{"podcastId":"1678378239","image":"https://is2-ssl.mzstatic.com/image/thumb/Podcasts116/v4/f6/cf/66/f6cf66f5-8696-1903-f3b3-e95c507c8e9c/mza_5620205623101925377.jpg/170x170bb.png","artist":"Cadence13 and Parallel","title":"This Little Light"}
    </div>
    <div>
      Channel{"podcastId":"1096830182","image":"https://is2-ssl.mzstatic.com/image/thumb/Podcasts126/v4/1f/ad/c7/1fadc7d4-1e22-beaa-720a-2e2988dc1521/mza_16091018887573148747.jpg/170x170bb.png","artist":"Interval Presents","title":"Drink Champs"}
    </div>
    <div>
      Channel{"podcastId":"1635211340","image":"https://is1-ssl.mzstatic.com/image/thumb/Podcasts112/v4/f4/53/be/f453be7f-2a34-0617-7bc1-bc57d9b747fc/mza_6548891788110664288.jpg/170x170bb.png","artist":"The Ringer","title":"60 Songs That Explain the '90s"}
    </div>
    <div>
      Channel{"podcastId":"1460157002","image":"https://is2-ssl.mzstatic.com/image/thumb/Podcasts125/v4/7b/cf/f6/7bcff6bb-5f99-6c2f-c6c5-3a9799f3df21/mza_8544742664200824246.jpg/170x170bb.png","artist":"Barstool Sports","title":"Million Dollaz Worth Of Game"}
    </div>
    <div>
      Channel{"podcastId":"1572182022","image":"https://is4-ssl.mzstatic.com/image/thumb/Podcasts116/v4/92/65/c6/9265c6dd-7db1-8160-c97c-bef564d23610/mza_16535014723154009651.jpg/170x170bb.png","artist":"Rory Farrell & Jamil \\"Mal\\" Clay","title":"New Rory & MAL"}
    </div>
    <div>
      Channel{"podcastId":"1628914491","image":"https://is1-ssl.mzstatic.com/image/thumb/Podcasts116/v4/2f/6e/cf/2f6ecf54-c735-97f2-d871-d46e2e909f5f/mza_11130508551645291638.jpg/170x170bb.png","artist":"REVOLT","title":"Caresha Please"}
    </div>
    <div>
      Channel{"podcastId":"1311004083","image":"https://is4-ssl.mzstatic.com/image/thumb/Podcasts112/v4/7b/12/d5/7b12d5ec-7264-6693-8a8e-e6b414a783c3/mza_10388338206053029687.jpg/170x170bb.png","artist":"Pushkin Industries","title":"Broken Record with Rick Rubin, Malcolm Gladwell, Bruce Headlam and Justin Richmond"}
    </div>
  </section>
</DocumentFragment>
`;

exports[`Search unit test should match snapshot when isFetching shows skeleton 1`] = `
<DocumentFragment>
  <div>
    SearchSkeleton
  </div>
</DocumentFragment>
`;
